activerecord/lib/active_record/associations/association.rb

```diff
@@ -211,7 +211,7 @@ def foreign_key_present?
         # the kind of the class of the associated objects. Meant to be used as
         # a sanity check when you are about to assign an associated record.
         def raise_on_type_mismatch!(record)
-          unless record.is_a?(reflection.klass) || record.is_a?(reflection.class_name.constantize)
+          unless record.is_a?(reflection.klass) || (record.is_a?(reflection.class_name.constantize) rescue false)
```

Comment:
Sorry, do you mean this?:

    record.is_a?(reflection.class_name.safe_constantize)

This breaks the tests because `#safe_constantize` returns `nil` when couldn't find a match and `record.is_a?(nil)` will thrown `TypeError: class or module required`.